File: code/Two Sum/1_twoSum.js
const nums = [1, 3, 7, 9, 2];

// const findTwoSum = (nums, target) => {
//   for (let p1 = 0; p1 < nums.length; p1++) {
//     const numberToFind = target - nums[p1];
//     for (let p2 = p1+1; p2 < nums.length; p2++) {
//       if (numberToFind === nums[p2]) {
//         return [p1, p2];
//       }
//     }
//   }
//   return null;
// };

const twoSum = (nums, target) => {
  const numsMap = {};
  for (let p = 0; p < nums.length; p++) {
    const currentMapVal = numsMap[nums[p]];
    if (currentMapVal >= 0) {
      return [currentMapVal, p];
    } else {
      numsMap[target - nums[p]] = p;
    }
  }
  return null;
};

console.log(twoSum(nums, 11));

File: code/Two Sum/1_two_sum.cpp
#include <vector>
#include <unordered_map>

using std::vector;
using std::unordered_map;

class Solution
{
public:
  vector<int> twoSum(vector<int> &nums, int target)
  {
    unordered_map<int, int> numsMap;
    vector<int> ans;

    for (int p = 0; p < nums.size(); p++)
    {
      if (numsMap.count(nums[p]) == 1)
      {
        ans.push_back(numsMap[nums[p]]);
        ans.push_back(p);
        return ans;
      }
      else
      {
        numsMap[target - nums[p]] = p;
      }
    }
    return ans;
  }
};